src/validate-sound.c

```diff
@@ -0,0 +1,368 @@
+/*
+ * Copyright © 2018 Red Hat, Inc
+ * Copyright © 2024 GNOME Foundation Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authors:
+ *       Matthias Clasen <mclasen@redhat.com>
+ *       Julian Sparber <jsparber@gnome.org>
+ */
+
+/* This is based on src/validate-icon.c */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <gst/gst.h>
+#include <gst/pbutils/pbutils.h>
+#include <glib/gstdio.h>
+#include <unistd.h>
+
+#ifdef __FreeBSD__
+#define execvpe exect
+#endif
+
+#define SOUND_VALIDATOR_GROUP "Sound Validator"
+
+static int
+validate_sound (int input_fd)
+{
+  g_autoptr(GKeyFile) key_file = NULL;
+  g_autofree char *key_file_data = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GstDiscoverer) discoverer = NULL;
+  g_autoptr(GstDiscovererInfo) info = NULL;
+  g_autoptr(GstDiscovererStreamInfo) audio_info = NULL;
+  g_autoptr(GstDiscovererStreamInfo) stream_info = NULL;
+  g_autoptr(GstDiscovererStreamInfo) stream_next = NULL;
+  GstDiscovererResult result;
+  const gchar *format = NULL;
+  g_autofree gchar *uri = NULL;
+
+  gst_init (NULL, NULL);
+
+  discoverer = gst_discoverer_new (GST_SECOND, &error);
+
+  if (!discoverer)
+    {
+      g_printerr ("validate-sound: Failed to create gstreamer discoverer: %s\n", error->message);
+      return 1;
+    }
+
+  uri = g_strdup_printf ("file:///proc/self/fd/%d", input_fd);
+  info = gst_discoverer_discover_uri (discoverer, uri, &error);
+  result = gst_discoverer_info_get_result (info);
+
+  switch (result) {
```

Comment:
Please also adjust the switch body indentation:

```suggestion
  switch (result)
    {
```